### PR TITLE
Redistribute new root creds in passthrough mode

### PR DESCRIPTION
**Pull request by akhil-rane**
This PR ensures that the dependent target secrets are updated when the root
credentials change in passthrough mode, even if the credential request was
recently synced.

jira: https://issues.redhat.com/browse/CO-1104

### CODE REVIEW

File: pkg/operator/credentialsrequest/credentialsrequest_controller_test.go [hunk 1]
```diff
@@ -1325,6 +1325,55 @@ func TestCredentialsRequestReconcile(t *testing.T) {
...
+			validate: func(c client.Client, t *testing.T) {
+				targetSecret := getSecret(c)
+				if assert.NotNil(t, targetSecret) {
+					assert.Equal(t, testRootAWSAccessKeyID,
+						string(targetSecret.Data["aws_access_key_id"]))
+					assert.Equal(t, testRootAWSSecretAccessKey,
+						string(targetSecret.Data["aws_secret_access_key"]))
+				}
+				cr := getCR(c)
+				assert.True(t, cr.Status.Provisioned)
```
Comment (joelddiaz) [inline pkg/operator/credentialsrequest/credentialsrequest_controller_test.go:1374]: add an assert here that the lastSyncCloudCredsSecretResource version was updated as expected.

File: pkg/aws/actuator/actuator.go [hunk 2]
```diff
@@ -235,6 +236,15 @@ func (a *AWSActuator) needsUpdate(ctx context.Context, cr *minterv1.CredentialsR
 		}
 		if mode == operatorv1.CloudCredentialsModePassthrough {
 			logger.Debug("will not perform permissions simulation because operator in mode %q", mode)
+			cloudCredsSecret, err := a.GetCloudCredentialsSecret(ctx, cr)
+			if err != nil {
+				log.WithError(err).Debug("error retrieving cloud credentials secret")
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/aws/actuator/actuator.go]: that make sense. Addressed this

File: pkg/aws/actuator/actuator.go [hunk 3]
```diff
@@ -235,6 +236,15 @@ func (a *AWSActuator) needsUpdate(ctx context.Context, cr *minterv1.CredentialsR
 		}
 		if mode == operatorv1.CloudCredentialsModePassthrough {
 			logger.Debug("will not perform permissions simulation because operator in mode %q", mode)
+			cloudCredsSecret, err := a.GetCloudCredentialsSecret(ctx, cr)
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/aws/actuator/actuator.go]: I was unaware of the design thinking behind this. Addressed it.

File: pkg/azure/actuator.go [hunk 4]
```diff
@@ -119,6 +120,37 @@ func isAzureCredentials(providerSpec *runtime.RawExtension) (bool, error) {
 	return isAzure, nil
 }
 
+// needsUpdate will return whether the current credentials satisfy what's being requested
+// in the CredentialsRequest
+func (a *Actuator) needsUpdate(ctx context.Context, cr *minterv1.CredentialsRequest) (bool, error) {
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/azure/actuator.go:124]: I suppose you mean status.ServicePrincipalName here?

File: pkg/openstack/actuator.go [hunk 5]
```diff
@@ -95,7 +95,13 @@ func (a *OpenStackActuator) sync(ctx context.Context, cr *minterv1.CredentialsRe
 	logger := a.getLogger(cr)
 	logger.Debug("running sync")
 
-	clouds, err := a.getRootCloudCredentialsSecretData(ctx, logger)
+	cloudCredsSecret, err := a.GetCloudCredentialsSecret(context.TODO(), cr)
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/openstack/actuator.go]: Oops, my bad

File: pkg/azure/actuator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joelddiaz) [reply pkg/azure/actuator.go:124]: Yes, the servicePrincipalName field.

File: pkg/azure/actuator.go [hunk 6]
```diff
@@ -119,6 +119,40 @@ func isAzureCredentials(providerSpec *runtime.RawExtension) (bool, error) {
...
+	if err != nil {
+		return true, fmt.Errorf("unable to decode ProviderStatus: %v", err)
+	}
+
+	if azureStatus.ServicePrincipalName == "" {
+		// passthrough-specifc checks here
+
+		credentialsRootSecret, err := a.GetCredentialsRootSecret(ctx, cr)
+		if err != nil {
+			log.WithError(err).Debug("error retrieving cloud credentials secret")
```
Comment (joelddiaz) [inline pkg/azure/actuator.go:145]: return the error here as it makes no sense to be in passthrough mode and have the root cloud creds be missing

File: pkg/azure/actuator.go [hunk 7]
```diff
@@ -214,6 +248,21 @@ func (a *Actuator) sync(ctx context.Context, cr *minterv1.CredentialsRequest) er
 	logger := a.getLogger(cr)
 	logger.Debug("running sync")
 
+	// Should we update anything
+	needsUpdate, err := a.needsUpdate(ctx, cr)
+	if err != nil {
+		logger.WithError(err).Error("error determining whether a credentials update is needed")
+		return &actuatoriface.ActuatorError{
+			ErrReason: minterv1.CredentialsProvisionFailure,
+			Message:   fmt.Sprintf("error determining whether a credentials update is needed: %v", err),
```
Comment (joelddiaz) [inline pkg/azure/actuator.go:257]: IIRC, the message here can be used to populate conditions on the CredentialsRequest. Can the inclusion of the contents of the `err` variable put us in a hotloop updating conditions?

Comment (akhil-rane) [reply pkg/azure/actuator.go:257]: We are going into hotloop if error message changes everytime and we do not know what azure (or any other cloud) will return as error message. So yes, may be need to remove it from every actuator. @dgoodwin do you agree?

Comment (akhil-rane) [reply pkg/azure/actuator.go:257]: We do have complete error in the logs so it should not be a problem for the end user

Comment (dgoodwin) [reply pkg/azure/actuator.go:257]: Yeah probably too risky for them to throw a request ID or something in there. Lets keep the clusteroperator condition message more static.

**Comment by akhil-rane** [comment]
@joelddiaz addressed your review comments

**Review by joelddiaz — COMMENTED** [review]
/lgtm

**Comment by akhil-rane** [comment]
/hold
failing e2es, putting a hold for now

**Comment by dgoodwin** [comment]
```
: [sig-cluster-lifecycle] Cluster completes upgrade expand_less | 2h30m0s
-- | --
Cluster did not complete upgrade: timed out waiting for the condition: Could not update credentialsrequest "openshift-cloud-credential-operator/openshift-machine-api-azure" (123 of 602): the object is invalid, possibly due to local cluster configuration
```

Something's gotta be going on here @akhil-rane , does this ring any bells as to why this might be failing forever on upgrade?

**Comment by dgoodwin** [comment]
From the cluster-version-operator logs, which applies all the resources in the release image during an upgrade:

```
E0909 17:01:02.504859       1 task.go:81] error running apply for credentialsrequest "openshift-cloud-credential-operator/openshift-machine-api-azure" (123 of 602): CredentialsRequest.cloudcredential.openshift.io "openshift-machine-api-azure" is invalid: status.lastSyncCloudCredsSecretResourceVersion: Required value
E0909 17:01:19.097189       1 task.go:81] error running apply for credentialsrequest "openshift-cloud-credential-operator/openshift-machine-api-azure" (123 of 602): CredentialsRequest.cloudcredential.openshift.io "openshift-machine-api-azure" is invalid: status.lastSyncCloudCredsSecretResourceVersion: Required value
E0909 17:01:45.109906       1 task.go:81] error running apply for credentialsrequest "openshift-cloud-credential-operator/openshift-machine-api-azure" (123 of 602): CredentialsRequest.cloudcredential.openshift.io "openshift-machine-api-azure" is invalid: status.lastSyncCloudCredsSecretResourceVersion: Required value
```

**Review by joelddiaz — COMMENTED** [review]
/lgtm

**Comment by akhil-rane** [comment]
@joelddiaz resolved merge conflicts here. Need LGTM again.

**Review by joelddiaz — COMMENTED** [review]
:+1: lets figure out what is going on with the e2e before an lgtm

**Comment by akhil-rane** [comment]
This PR had cleared all the tests before so I thought may be its a flake

**Comment by joelddiaz** [comment]
/lgtm
fair enough, let's see if we can get through CI

**Pull request merged**